### PR TITLE
Implements sponsored new tab images

### PR DESCRIPTION
**Pull request by SergeyZhukovsky**
closes https://github.com/brave/browser-android-tabs/issues/2363

### CODE REVIEW

File: chrome/android/java/src/org/chromium/chrome/browser/NTPUpdaterWorker.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        @Override
+        public void run() {
+          for (;;) {
+              try {
+                  Thread.sleep(INTERVAL_TO_UPDATE);
+                  NTPUpdater.UpdateNTP(mContext);
+              }
+              catch(Exception exc) {
+                  // Just ignore it if we cannot update
+                  Log.i("TAG", "Update loop exception: " + exc);
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/NTPUpdaterWorker.java]: This should probably be logged as error

File: chrome/android/java/src/org/chromium/chrome/browser/init/ChromeBrowserInitializer.java [hunk 2]
```diff
@@ -190,6 +192,24 @@ protected Long doInBackground() {
...
 
+    // NTP update
+    class UpdateNTPAsyncTask extends AsyncTask<Long> {
+        @Override
+        protected Long doInBackground() {
+            try {
+                NTPUpdater.UpdateNTP(ContextUtils.getApplicationContext());
+            }
+            catch(Exception exc) {
+                // Just ignore it if we cannot update
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/ChromeBrowserInitializer.java]: Would be good to have log here

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+        return false;
+    }
+
+    public static boolean downloadFile(Context context, String urlString, String countryCode, String fileName, boolean isJSONFile) {
+        byte[] buffer = null;
+        InputStream inputStream = null;
+        HttpURLConnection connection = null;
+        boolean res = false;
+        try {
+            Log.i("NTP", "Downloading %s", fileName);
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Not sure if we need this log

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+            URL url = new URL(urlString);
+            connection = (HttpURLConnection) url.openConnection();
+            int length = connection.getContentLength();
+
+            if (isJSONFile) {
+                EtagObject previousEtag = getETagInfo(context, ETAG_PREPEND_NTP);
+                String etag = connection.getHeaderField("ETag");
+                Log.i("NTP", "etag : "+ etag);
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Not sure if we need this log

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+                if (etag.equals(previousEtag.mEtag)) {
+                    downloadFile = false;
+                } else {
+                    removePreviousFiles(context, countryCode);
+                }
+
+                previousEtag.mEtag = etag;
+                saveETagInfo(context, ETAG_PREPEND_NTP, previousEtag);
+                if (!downloadFile && new File(PathUtils.getDataDirectory(), countryCode + "_" + NTP_JSON_FILE_NAME).exists()) {
+                    Log.i("NTP", "do not downaload json", fileName);
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Not sure if we need this log. Probably should be logged as error.

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+                    parseJSONFile(context, countryCode);
+                    return false;
+                }
+            }
+
+            connection.connect();
+
+            if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
+                Log.i("NTP", "response code : "+ connection.getResponseCode());
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Not sure if we need this log. Probably should be logged as error.

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+            int n = - 1;
+            int totalReadSize = 0;
+            try {
+                while ((n = inputStream.read(buffer)) != -1) {
+                    outputStream.write(buffer, 0, n);
+                    totalReadSize += n;
+                }
+            }
+            catch (IllegalStateException exc) {
+                Log.i("NTP", "IllegalStateException");
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Not sure if we need this log. Probably should be logged as error with more info.

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+                // Sometimes it gives us that exception, found that we should do that way to avoid it:
+                // Each HttpURLConnection instance is used to make a single request but the
+                // underlying network connection to the HTTP server may be transparently shared by other instance.
+                // But we do that way, so just wrapped it for now and we will redownload the file on next request
+            }
+            outputStream.close();
+            if (length != totalReadSize || length != tempFile.length()) {
+                Log.i("NTP", "download failed");
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Probably should be logged as error.

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            } else {
+              // We downloaded the file with success, rename it now to .json
+              File renameTo = new File(PathUtils.getDataDirectory(), countryCode + "_" + fileName);
+              if (!tempFile.exists() || !tempFile.renameTo(renameTo)) {
+              }
+              res = true;
+              Log.i("NTP", "Downloaded %s", renameTo.getPath());
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Not sure if we need this log.

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+                    int focalPointY = !imageObj.isNull("focalPoint") ? imageObj.getJSONObject("focalPoint").getInt("y") : 0;
+
+                    SponsoredImage sponsoredImage = new SponsoredImage(
+                        imageObj.getString("imageUrl"), 
+                        focalPointX, 
+                        focalPointY);
+                    sponsoredImages.add(sponsoredImage);
+                }
+                downloadImages(context, countryCode, sponsoredImages);
+                Log.i("NTP", "JSON obj : "+mainObj.toString());
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Not sure if we need this log.

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+            }
+        } catch (Exception e) {
+            Log.i("NTP", e.getMessage());
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Probably should be logged as error.

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+            BufferedReader reader = new BufferedReader(new InputStreamReader(is));
+            StringBuilder sb = new StringBuilder();
+            String line = null;
+            while ((line = reader.readLine()) != null) {
+              sb.append(line);
+            }
+            reader.close();
+            return sb.toString();
+        } catch (Exception e) {
+            Log.i("NTP", e.getMessage());
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Probably should be logged as error.

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+        if (null == dataDirPath) {
+            return;
+        }
+        File[] fileList = dataDirPath.listFiles();
+
+        for (File file : fileList) {
+            String filePath = file.getPath();
+            String sFileName = filePath.substring(filePath.lastIndexOf(File.separator)+1);
+            if (sFileName.startsWith(countryCode + "_")) {
+                Log.i("NTP", "Deleting : " + sFileName);
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: Not sure if we need this log.

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPUpdater.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+                // Don't set an update flag on initial download
+                ContextUtils.getAppSharedPreferences().edit()
+                  .putBoolean(UPDATE_NTP, true)
+                  .apply();
+              }
+            } finally {
+                mAvailable.release();
+            }
+        } catch (InterruptedException exc) {
+            Log.i("NTP", "InterruptedException");
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/init/NTPUpdater.java]: Should be logged as error.

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java [hunk 15]
```diff
@@ -491,24 +530,98 @@ public SnapScrollHelper getSnapScrollHelper() {
...
+
+                    Bitmap imageBitmap = null;
+                    float imageWidth;
+                    float imageHeight;
+                    float centerPoint;
+
+                    if (ntpImage instanceof SponsoredImage) {
+                        SponsoredImage sponsoredImage = (SponsoredImage) ntpImage;
+                        File imageFile = new File(PathUtils.getDataDirectory(), countryCode + "_" + sponsoredImage.getImageUrl());
+                        Log.i("NTP", " imageFile : " + imageFile.getPath());
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java]: Not sure if we need this log.

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java [hunk 16]
```diff
@@ -491,24 +530,98 @@ public SnapScrollHelper getSnapScrollHelper() {
...
+                        try {
+                            Uri imageFileUri = Uri.parse("file://"+imageFile.getAbsolutePath());
+                            Log.i("NTP", " imageFileUri : " + imageFileUri.getPath());
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java]: Not sure if we need this log.

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java [hunk 17]
```diff
@@ -491,24 +530,98 @@ public SnapScrollHelper getSnapScrollHelper() {
...
+                            InputStream inputStream = mTab.getActivity().getContentResolver().openInputStream(imageFileUri);
+                            imageBitmap = BitmapFactory.decodeStream(inputStream);
+                        } catch (Exception exc) {
+                            Log.i("NTP", exc.getMessage());
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java]: Should be logged as error.

File: chrome/android/java/strings/android_chrome_strings.grd [hunk 18]
```diff
@@ -2164,7 +2164,7 @@ To change this setting, <ph name="BEGIN_LINK">&lt;resetlink&gt;</ph>reset sync<p
         Sans Serif
       </message>
       <message name="IDS_SERIF" desc="Title of the selector button that will change the typeface of distilled pages to use a Serif font.">
-        Serif
+chrome/android/java/strings/android_chrome_strings.grd        Serif
```
Comment (samartnik) [inline chrome/android/java/strings/android_chrome_strings.grd]: What is this change for?

File: chrome/android/java/src/org/chromium/chrome/browser/NTPUpdaterWorker.java (hunk 1 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/NTPUpdaterWorker.java]: fixed

File: chrome/android/java/src/org/chromium/chrome/browser/init/ChromeBrowserInitializer.java (hunk 2 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/ChromeBrowserInitializer.java]: added

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 3 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: removed

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 4 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: removed

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 5 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: removed

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 6 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: changed to an error

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 7 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: modified to an error

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 8 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: modified to an error

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 9 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: removed

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 10 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: removed

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 11 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: modified to an error

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 12 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: modified to an error

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java (hunk 13 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPDownloadUtils.java]: removed

File: chrome/android/java/src/org/chromium/chrome/browser/init/NTPUpdater.java (hunk 14 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/init/NTPUpdater.java]: modified to an error

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java (hunk 15 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java]: removed

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java (hunk 16 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java]: removed

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java (hunk 17 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java]: modified to an error

File: chrome/android/java/strings/android_chrome_strings.grd (hunk 18 above)
Comment (SergeyZhukovsky) [reply chrome/android/java/strings/android_chrome_strings.grd]: not sure either, reverted it

File: chrome/android/java/src/org/chromium/chrome/browser/NTPUpdaterWorker.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+package org.chromium.chrome.browser;
+
+import android.content.Context;
+
+import org.chromium.base.Log;
+import org.chromium.chrome.browser.init.NTPUpdater;
+
+public class NTPUpdaterWorker {
+
+    private static final int INTERVAL_TO_UPDATE = 1000 * 60 * 120;    // Milliseconds
```
Comment (diracdeltas) [inline chrome/android/java/src/org/chromium/chrome/browser/NTPUpdaterWorker.java]: i thought this was supposed to be 5 hours, not 2?

can we just do what is done on iOS, including the randomness? https://github.com/brave/brave-ios/pull/2216/files#diff-f23b4aaeb8eacb563731bbc6d97d142aR135

File: chrome/android/java/res/xml/qa_preferences.xml [hunk 20]
```diff
@@ -17,5 +17,12 @@
         android:summaryOn="@string/text_on"
         android:summaryOff="@string/text_off"
         android:defaultValue="true" />
+
+    <org.chromium.chrome.browser.preferences.ChromeSwitchPreference
+        android:key="qa_debug_ntp"
+        android:title="Enable debugging NTP"
```
Comment (samartnik) [inline chrome/android/java/res/xml/qa_preferences.xml]: Do we need to translate this string?

File: chrome/android/java/src/org/chromium/chrome/browser/NTPUpdaterWorker.java [hunk 21]
```diff
@@ -36,18 +43,29 @@ public void Stop() {
...
+    }
+
     class UpdateThread extends Thread {
 
         @Override
         public void run() {
           for (;;) {
               try {
-                  Thread.sleep(INTERVAL_TO_UPDATE);
+                  Log.e("NTP", "Time interval to update NTP : " + getUpdateInterval());
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/NTPUpdaterWorker.java]: These logs seems to be excessive

File: chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabCreatorManager.java [hunk 22]
```diff
@@ -90,7 +103,26 @@ public final boolean createTabWithWebContents(
...
+                            chromeTabbedActivity.openNewOrSelectExistingTab(UrlConstants.NTP_URL);
+                            chromeTabbedActivity.hideOverviewTemp();
+                        }
+                    } else {
+                        launchUrl(UrlConstants.NTP_URL, TabLaunchType.FROM_CHROME_UI);
+                    }
+                } else {
+                    launchUrl(UrlConstants.NTP_URL, TabLaunchType.FROM_CHROME_UI);
+                }
+                Log.i("NTP", "Tab opened");
```
Comment (samartnik) [inline chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabCreatorManager.java]: Seems to be unnecessary log

File: chrome/android/java/res/xml/qa_preferences.xml (hunk 20 above)
Comment (samartnik) [reply chrome/android/java/res/xml/qa_preferences.xml]: @SergeyZhukovsky pointed out that it is in QA section, so we don't need translation

File: chrome/android/java/src/org/chromium/chrome/browser/NTPUpdaterWorker.java (hunk 21 above)
Comment (deeppandya) [reply chrome/android/java/src/org/chromium/chrome/browser/NTPUpdaterWorker.java]: my bad. updating it.

File: chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabCreatorManager.java (hunk 22 above)
Comment (deeppandya) [reply chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabCreatorManager.java]: my bad. updating it.

File: chrome/android/java/res/xml/qa_preferences.xml (hunk 20 above)
Comment (deeppandya) [reply chrome/android/java/res/xml/qa_preferences.xml]: it's in QA preferences so don't need translation

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageLayout.java [hunk 23]
```diff
@@ -794,7 +797,16 @@ private void updateTileGridPlaceholderVisibility() {
                 ViewStub placeholderStub = findViewById(R.id.tile_grid_placeholder_stub);
                 mTileGridPlaceholder = placeholderStub.inflate();
                 SharedPreferences sharedPreferences = ContextUtils.getAppSharedPreferences();
-                if(sharedPreferences.getBoolean(BackgroundImagesPreferences.PREF_SHOW_BACKGROUND_IMAGES, true)) {
+
+                boolean isMoreTabs = false;
+                ChromeTabbedActivity chromeTabbedActivity = BraveRewardsHelper.getChromeTabbedActivity();
+                if(chromeTabbedActivity != null) {
+                    TabModel tabModel = chromeTabbedActivity.getCurrentTabModel();
+                    isMoreTabs = tabModel.getCount() > 10 ? true : false;
```
Comment (SergeyZhukovsky) [inline chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageLayout.java]: Probably better to put 10 on the top of class declaration as a static final member for easier modifications in the future

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageLayout.java [hunk 24]
```diff
@@ -809,7 +821,14 @@ private void updateTileGridPlaceholderVisibility() {
 
     private int getMaxTileRows() {
         SharedPreferences sharedPreferences = ContextUtils.getAppSharedPreferences();
-        if(sharedPreferences.getBoolean(BackgroundImagesPreferences.PREF_SHOW_BACKGROUND_IMAGES, true)) {
+        boolean isMoreTabs = false;
+        ChromeTabbedActivity chromeTabbedActivity = BraveRewardsHelper.getChromeTabbedActivity();
+        if(chromeTabbedActivity != null) {
+            TabModel tabModel = chromeTabbedActivity.getCurrentTabModel();
+            isMoreTabs = tabModel.getCount() >= 10 ? true : false;
```
Comment (SergeyZhukovsky) [inline chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageLayout.java]: the same as above about 10

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java [hunk 25]
```diff
@@ -185,10 +189,21 @@ public NewTabPageView(Context context, AttributeSet attrs) {
...
     public void onConfigurationChanged(Configuration newConfig) {
-        NTPImage ntpImage = mTab.getTabNTPImage();
-        checkForNonDistruptiveBanner(ntpImage);
-        super.onConfigurationChanged(newConfig);
-        showNTPImage(ntpImage);
+        boolean isMoreTabs = false;
+        ChromeTabbedActivity chromeTabbedActivity = BraveRewardsHelper.getChromeTabbedActivity();
+        if(chromeTabbedActivity != null) {
+            TabModel tabModel = chromeTabbedActivity.getCurrentTabModel();
+            isMoreTabs = tabModel.getCount() > 10 ? true : false;
```
Comment (SergeyZhukovsky) [inline chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java]: we can take 10 from a NewTabPageLayout if do a public method there, so we have only one place where we define it

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java [hunk 26]
```diff
@@ -331,6 +344,25 @@ public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
...
         manager.addDestructionObserver(NewTabPageView.this::onDestroy);
 
+        boolean isMoreTabs = false;
+
+        for (Activity ref : ApplicationStatus.getRunningActivities()) {
+            if (!(ref instanceof ChromeTabbedActivity)) continue;
+            ChromeTabbedActivity chromeTabbedActivity =  (ChromeTabbedActivity)ref;
+            if(chromeTabbedActivity != null) {
+                TabModel tabModel = chromeTabbedActivity.getCurrentTabModel();
+                isMoreTabs = tabModel.getCount() > 10 ? true : false;
```
Comment (SergeyZhukovsky) [inline chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java]: the same

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java [hunk 27]
```diff
@@ -381,7 +413,15 @@ private void updateBraveStats() {
         TextView httpsUpgradesTextView = (TextView) mBraveStatsView.findViewById(R.id.brave_stats_text_https);
         TextView estTimeSavedTextView = (TextView) mBraveStatsView.findViewById(R.id.brave_stats_text_time);
 
-        if(mSharedPreferences.getBoolean(BackgroundImagesPreferences.PREF_SHOW_BACKGROUND_IMAGES, true)) {
+        boolean isMoreTabs = false;
+        ChromeTabbedActivity chromeTabbedActivity = BraveRewardsHelper.getChromeTabbedActivity();
+        if(chromeTabbedActivity != null) {
+            TabModel tabModel = chromeTabbedActivity.getCurrentTabModel();
+            isMoreTabs = tabModel.getCount() > 10 ? true : false;
```
Comment (SergeyZhukovsky) [inline chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java]: 10

File: chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java [hunk 28]
```diff
@@ -532,7 +572,16 @@ public SnapScrollHelper getSnapScrollHelper() {
 
     private void showNTPImage(NTPImage ntpImage) {
         updateOrientedUI();
-        if(mSharedPreferences.getBoolean(BackgroundImagesPreferences.PREF_SHOW_BACKGROUND_IMAGES, true)) {
+
+        boolean isMoreTabs = false;
+        ChromeTabbedActivity chromeTabbedActivity = BraveRewardsHelper.getChromeTabbedActivity();
+        if(chromeTabbedActivity != null) {
+            TabModel tabModel = chromeTabbedActivity.getCurrentTabModel();
+            isMoreTabs = tabModel.getCount() > 10 ? true : false;
```
Comment (SergeyZhukovsky) [inline chrome/android/java/src/org/chromium/chrome/browser/ntp/NewTabPageView.java]: 10

File: chrome/android/java/src/org/chromium/chrome/browser/tab/Tab.java [hunk 29]
```diff
@@ -349,9 +353,18 @@ public void onViewDetachedFromWindow(View view) {
...
         mFingerprintsBlocked = 0;
         
-        ntpImage = getNTPImage();
-        index = SponsoredImageUtil.imageIndex;
-        updateBannerPref();
+        boolean isMoreTabs = false;
+        ChromeTabbedActivity chromeTabbedActivity = BraveRewardsHelper.getChromeTabbedActivity();
+        if(chromeTabbedActivity != null) {
+            TabModel tabModel = chromeTabbedActivity.getCurrentTabModel();
+            isMoreTabs = tabModel.getCount() > 10 ? true : false;
```
Comment (SergeyZhukovsky) [inline chrome/android/java/src/org/chromium/chrome/browser/tab/Tab.java]: 10

File: chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabCreatorManager.java [hunk 30]
```diff
@@ -90,7 +103,26 @@ public final boolean createTabWithWebContents(
...
             try {
                 TraceEvent.begin("TabCreator.launchNTP");
-                launchUrl(UrlConstants.NTP_URL, TabLaunchType.FROM_CHROME_UI);
+
+                SharedPreferences mSharedPreferences = ContextUtils.getAppSharedPreferences();
+
+                ChromeTabbedActivity chromeTabbedActivity = BraveRewardsHelper.getChromeTabbedActivity();
+                if(chromeTabbedActivity != null && Build.VERSION.SDK_INT <= Build.VERSION_CODES.M) {
+                    TabModel tabModel = chromeTabbedActivity.getCurrentTabModel();
+                    if (tabModel.getCount() >= 10 && mSharedPreferences.getBoolean(BackgroundImagesPreferences.PREF_SHOW_BACKGROUND_IMAGES, true)) {
```
Comment (SergeyZhukovsky) [inline chrome/android/java/src/org/chromium/chrome/browser/tabmodel/TabCreatorManager.java]: We should really have that 10 in one place only. In some helper probably

File: chrome/android/java/src/org/chromium/chrome/browser/widget/tile/TileWithTextView.java [hunk 31]
```diff
@@ -59,9 +62,17 @@ public void setTitle(String title, int titleLines) {
...
 
-        ChromeTabbedActivity chromeTabbedActivity = BraveRewardsHelper.getChromeTabbedActivity();
         SharedPreferences mSharedPreferences = ContextUtils.getAppSharedPreferences();
-        if(mSharedPreferences.getBoolean(BackgroundImagesPreferences.PREF_SHOW_BACKGROUND_IMAGES, true)) {
+
+        boolean isMoreTabs = false;
+        ChromeTabbedActivity chromeTabbedActivity = BraveRewardsHelper.getChromeTabbedActivity();
+        if(chromeTabbedActivity != null) {
+            TabModel tabModel = chromeTabbedActivity.getCurrentTabModel();
+            isMoreTabs = tabModel.getCount() >= 10 ? true : false;
```
Comment (SergeyZhukovsky) [inline chrome/android/java/src/org/chromium/chrome/browser/widget/tile/TileWithTextView.java]: 10

**Comment by SergeyZhukovsky** [comment]
it looks good to me, I cannot approve it as I was the PR creator :). @diracdeltas does that logic of updates look good to you https://github.com/brave/browser-android-tabs/pull/2490/files#diff-b26f522e71d45d7530b603abdc220e4aR19.

**Pull request merged**